# Week 4 Assignment

In the following questions, we will be working with three bonds:

- B1 is a 15 Year Bond with a Face Value of $1000 that pays a 5% coupon semi-annually (2 times a year)
- B2 is a 5 Year Bond with a Face value of $1000 that pays a 6% coupon quarterly (4 times a year)
- B3 is a 10 Year Zero-Coupon Bond with a Face Value of $1000 (Hint: you can still use the erk.bond_cash_flows() and erk.bond_price() by setting the coupon amount to 0% and coupons_per_year to 1)

Assume the yield curve is flat at 5%. Duration refers to Macaulay Duration (Hint: the macaulay_duration function gives as output the duration expressed in periods and not in years. If you want to get the yearly duration you need to divide the duration for coupons_per_year; e.g.: duration_B2 = erk.macaulay_duration(flows_B2, 0.05/4)/4).

In [1]:
import numpy as np
import pandas as pd
import edhec_risk_kit_129 as erk

## 1. Question 1
Which of the three bonds is the most expensive?

## 2. Question 2
Which of the three bonds is the least expensive?

## 3. Question 3
What is the price of the 10 Year Zero Coupon Bond B3?

In [8]:
# yield curve (discount rate)
yc=.05
# b1 = maturity=15 years, value=1000, coupon rate=.05, coupons per year=2
b1_price = erk.bond_price(15, 1000, .05, 2, yc)
print(np.round(b1_price))
# b2 = maturity=5 years, value=1000, coupon rate=.06, coupons per year=4
b2_price = erk.bond_price(5, 1000, .06, 4, yc)
print(np.round(b2_price))
# b3 = maturity=10 years, value=1000, coupon rate=0, coupons per year=1 (see hint)
b3_price = erk.bond_price(10, 1000, 0, 1, yc)
print(np.round(b3_price))

0    1000.0
dtype: float64
0    1044.0
dtype: float64
0    614.0
dtype: float64


0    613.913254
dtype: float64

## 4. Question 4
Which of the three bonds has the highest (Macaulay) Duration?

## 5. Question 5
Which of the three bonds has the lowest (Macaulay) Duration?

## 6. Question 6
What is the duration of the 5 year bond B2?

In [10]:
# b1 = maturity=15 years, value=1000, coupon rate=.05, coupons per year=2
b1_flows = erk.bond_cash_flows(15, 1000, .05, 2)
b1_duration = erk.macaulay_duration(b1_flows, yc/2)/2
print(np.round(b1_duration, 2))
# b2 = maturity=5 years, value=1000, coupon rate=.06, coupons per year=4
b2_flows = erk.bond_cash_flows(5, 1000, .06, 4)
b2_duration = erk.macaulay_duration(b2_flows, yc/4)/4
print(np.round(b2_duration, 2))
# b3 = maturity=10 years, value=1000, coupon rate=0, coupons per year=1 (see hint)
b3_flows = erk.bond_cash_flows(10, 1000, 0, 1)
b3_duration = erk.macaulay_duration(b3_flows, yc)
print(np.round(b3_duration, 2))

10.73
4.37
10.0


## 7. Question 7
Assume a sequence of 3 liabilities of $100,000, $200,000 and $300,000 that are 3, 5 and 10 years away, respectively. What is the Duration of the liabilities?
(Reminder: Assume the yield curve is flat at 5%. Duration refers to Macaulay Duration)

In [11]:
liabilities = pd.Series(data = [100000, 200000, 300000], index=[3, 5, 10])
l_duration = erk.macaulay_duration(liabilities, yc)
l_duration

6.750917852744651

## 8. Question 8
Assuming the same set of liabilities as the previous question (i.e. a sequence of 3 liabilities of $100,000, $200,000 and $300,000 that are 3, 5 and 10 years away, respectively) build a Duration Matched Portfolio of B1 and B2 to match these liabilities. What is the weight of B2 in the portfolio? (Hint: the code we developed in class erk.match_durations() assumes that all the bonds have the same number of coupons per year. This is not the case here, so you will either need to enhance the code or compute the weight directly, e.g. by entering the steps in a Jupyter Notebook Cell or at the Python Command Line)

In [13]:
d_l=b1_duration
d_s=b2_duration
d_t=l_duration
b2_weight=(d_l - d_t)/(d_l - d_s)
b2_weight

0.6257830075060314

## 9. Question 9
Assume you can use any of the bonds B1, B2 and B3 to build a duration matched bond portfolio matched to the liabilities. Which combination of 2 bonds can you NOT use to build a duration matched bond portfolio?
- B1+B2
- **B1+B3** _(because both maturities exceed 6.75)_
- ANY PAIR WILL WORK
- B2+B3

## 10. Question 10
Assuming the same liabilities as the previous questions (i.e. a sequence of 3 liabilities of $100,000, $200,000 and $300,000 that are 3, 5 and 10 years away, respectively), build a Duration Matched Portfolio of B2 and B3 to match the liabilities. What is the weight of B2 in this portfolio?

In [14]:
d_l=b3_duration
d_s=b2_duration
d_t=l_duration
b2_weight=(d_l - d_t)/(d_l - d_s)
b2_weight

0.5774465770256697